routesrv/routesrv.go

```diff
@@ -0,0 +1,154 @@
+		IngressClass:                      opts.KubernetesIngressClass,
+		RouteGroupClass:                   opts.KubernetesRouteGroupClass,
+		ReverseSourcePredicate:            opts.ReverseSourcePredicate,
+		WhitelistedHealthCheckCIDR:        opts.WhitelistedHealthCheckCIDR,
+		PathMode:                          opts.KubernetesPathMode,
+		KubernetesNamespace:               opts.KubernetesNamespace,
+		KubernetesEastWestRangeDomains:    opts.KubernetesEastWestRangeDomains,
+		KubernetesEastWestRangePredicates: opts.KubernetesEastWestRangePredicates,
+		DefaultFiltersDir:                 opts.DefaultFiltersDir,
+		BackendNameTracingTag:             opts.OpenTracingBackendNameTag,
+		OnlyAllowedExternalNames:          opts.KubernetesOnlyAllowedExternalNames,
+		AllowedExternalNames:              opts.KubernetesAllowedExternalNames,
+	})
+	if err != nil {
+		return nil, err
+	}
+	rs.poller = &poller{
+		client:  dataclient,
+		timeout: opts.SourcePollTimeout,
+		b:       b,
+		quit:    make(chan struct{}),
+		tracer:  tracer,
+	}
+
+	rs.wg = &sync.WaitGroup{}
+
+	return rs, nil
+}
+
+// StartUpdates starts the data source polling process.
+func (rs *RouteServer) StartUpdates() {
+	rs.wg.Add(1)
+	go rs.poller.poll(rs.wg)
+}
+
+// StopUpdates stop the data source polling process.
+func (rs *RouteServer) StopUpdates() {
+	rs.poller.quit <- struct{}{}
+}
+
+// ServeHTTP serves kept eskip-formatted routes under /routes
+// endpoint. Additionally it provides a simple health check under
+// /health and Prometheus-compatible metrics under /metrics.
+func (rs *RouteServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	rs.server.Handler.ServeHTTP(w, r)
+}
+
+func newShutdownFunc(rs *RouteServer) func(delay time.Duration) {
+	once := &sync.Once{}
+	rs.wg.Add(1)
+
+	return func(delay time.Duration) {
+		once.Do(func() {
+			defer rs.wg.Done()
+			defer rs.StopUpdates()
+
+			log.Infof("shutting down the server in %s...", delay)
+			time.Sleep(delay)
+			if err := rs.server.Shutdown(context.Background()); err != nil {
+				log.Error("unable to shut down the server: ", err)
+			}
+			log.Info("server shut down")
+		})
+	}
+}
+
+// Run starts a route server set up according to the passed options.
+// It is a blocking call designed to be used as a single call/entry point,
+// when running the route server as a standalone binary. It returns, when
+// the server is closed, which can happen due to server startup errors or
+// gracefully handled SIGTERM signal. In case of a server startup error,
+// the error is returned as is.
+func Run(opts Options) error {
+	rs, err := New(opts)
+	if err != nil {
+		return err
+	}
+
+	shutdown := newShutdownFunc(rs)
+
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGTERM)
+	go func() {
+		<-sigs
+		shutdown(opts.WaitForHealthcheckInterval)
+	}()
+
+	rs.StartUpdates()
+
+	if err = rs.server.ListenAndServe(); err != http.ErrServerClosed {
+		go shutdown(0)
```

Comment:
In order we want to run in production in Kubernetes, I think we need to change this to use `opt.ShutdownDelay` a `time.Duration` we set via flag.

Reply:
In my understanding there is no need for that. If the error received **is not** `http.ErrServerClosed`, it would suggest it is a startup error and thus the server was never started. I think it is just better to fail fast in this kind of scenario. Otherwise we would just wait for `opt.ShutdownDelay` without a need, thus delaying detection of a potential configuration failure for example.
All of this reasoning would be invalid, if other errors can be returned **after** the server was started. Maybe I am missing something, is there a possibility for other error than `http.ErrServerClosed` after the server is properly started?